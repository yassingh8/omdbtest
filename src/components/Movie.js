import React from 'react';
import axios from 'axios';
//import {Route, Switch } from 'react-router-dom';
import {connect} from 'react-redux';
import { listAction, showList } from '../actions/addToList';
import {
  Card,
  Button,
} from 'react-bootstrap';

class Movie extends React.Component {
    constructor(props) {
      super(props);
      // this.data = {};
    }

    getData = () => {
      return axios
      .get(
          'https://www.omdbapi.com/?apikey=d0ecb0ed&i='+this.props.movieID+'&plot=full')
      .then(res => {
          this.data = res.data;
          console.log(this.data);
      });
    }
    
    handleClick(i) {
      var l = this.props.myList
      const j = this.props.myList.includes(i);
      if (!j) {
        if (l.length<10) {
          console.log("add clicked! "+ i)
          if (l.indexOf(i)<0) {
            l = l.concat([i])
            this.props.listAction(l);
          }
        }
        else alert("List full...delete some items")
      }
      else {
        console.log("delete clicked!")
        l.splice(l.indexOf(i),1)
        this.props.listAction(l);
      }   
    }
  
    // componentWillMount() {
    //   console.log('card didmount!');
    //   axios
    //   .get(
    //       'https://www.omdbapi.com/?apikey=d0ecb0ed&i='+this.props.movieID+'&plot=full')
    //   .then(res => {
    //       this.data = res.data;
    //       console.log(this.data);
    //   });
    // }
  
    render() {
      console.log('card render');
      
      const {
        Title,
        Poster,
        Year
      } = this.props.data;
      // } = this.getData();
      
    console.log(this.props.data + ' data');

    if (!Poster || Poster === 'N/A') {
        return null;
    }
      return(
            <Card style={{ width: "15rem"}} border="dark">
                <Card.Img variant="top" src={Poster}/>
                <Card.Body>      
                <Card.Title>{Title}</Card.Title>
                    <Card.Subtitle className="mb-2 text-muted">Year</Card.Subtitle>
                    <Card.Text>{Year}</Card.Text>
                </Card.Body>
                <Card.Footer>
                    <Button size="sm" onClick={()=>{return this.handleClick(this.props.data)}}>
                        {this.props.myList.includes(this.props.data) ? "Delete" : "Add"}
                    </Button>
                </Card.Footer>     
            </Card>
      );
    }
}

const mapStateToProps = state => ({
  searchTerm: state.favourite.searchTerm,
  myList: state.favourite.myList,
  movieList: state.favourite.movieList
})

export default connect(mapStateToProps, { listAction, showList })(Movie);