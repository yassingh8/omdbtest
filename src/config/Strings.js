export const SEARCH_MOVIE = 'SEARCH_MOVIE';
export const FETCH_MOVIES = 'FETCH_MOVIES';
export const BUTTON_CLICK = 'BUTTON_CLICK';
export const SHOW_LIST = 'SHOW_LIST';
export const SHOW_HOME = 'SHOW_HOME';