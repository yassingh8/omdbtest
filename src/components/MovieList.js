import React from 'react';
// import { useSelector } from 'react-redux';
// import { createSelector } from 'reselect';
import axios from 'axios';
//import {Route, Switch } from 'react-router-dom';
import { useSelector, useDispatch} from 'react-redux';
import { fetchMovies, searchMovie, showList, showHome } from '../actions/addToList';
import {
  Form,
  FormControl,
  Nav,
  Navbar,
  Button,
  CardColumns,
  Container
} from 'react-bootstrap';
import { MdSearch, MdLibraryAdd } from 'react-icons/md';
import Movie from './Movie';


// const getProps = createSelector(
//     state => state.favourite,
//     todos => todos.filter(todo => todo.isDone).length
//   )

// class MoviesList extends React.Component {
const MoviesList = (props) => { 
    // state = {
    //     moviesList: ['tt2294629'],
    //     searchTerm: ''
    // };
    const favourite = useSelector(state => state.favourite);
    const dispatch = useDispatch();

    const search = event => {
        event.preventDefault();
        axios.get(`https://www.omdbapi.com/?apikey=d0ecb0ed&s=${favourite.searchTerm}&plot=full`)
        .then(res => dispatch(fetchMovies(res.data.Search)))
        .catch(err => console.log(err));
    };

    const handleChange = event => {
        dispatch(searchMovie(event.target.value));
        // setState({
        //     searchTerm: event.target.value
        // });
    };

    const showFav = () => {
        dispatch(showList(favourite.myList));
    }

    const goHome = () => {
        dispatch(showHome());
    }
    // render() {
        // const { moviesList } = state;

        return (
            <Container>
                <Navbar bg="light" expand="lg">
                    <Navbar.Brand onClick={goHome}>Search Movies</Navbar.Brand>
                    <Navbar.Collapse id="basic-navbar-nav">
                        <Nav className="mr-auto">
                            <Nav.Link onClick={showFav}>
                                <MdLibraryAdd/>
                                My List
                            </Nav.Link>
                        </Nav>
                        <Form inline>
                            <FormControl type="text" placeholder="Search for a movie" onChange={handleChange} className="mr-sm-2" />
                            <Button variant="outline-success" onClick={search}>
                                <MdSearch/>
                            </Button>
                        </Form>
                    </Navbar.Collapse>
                </Navbar>
                <br/>
                <CardColumns>
                    {favourite.movieList && favourite.movieList.length > 0 ? (
                        favourite.movieList.map(movie => (
                            <Movie data={movie} key={movie} />
                        ))
                    ) : (
                        <p>
                            Couldn't find any movie. Please search again using
                            another search criteria.
                        </p>
                    )}
                </CardColumns>

            </Container>
        );
    // }
}

// const mapStateToProps = state => ({
//     searchTerm: state.favourite.searchTerm,
//     myList: state.favourite.myList,
//     movieList: state.favourite.movieList,
//     showMy: state.favourite.showMy
//   })
  
// export default connect(mapStateToProps, { searchMovie, showList, showHome, fetchMovies })(MoviesList);
export default MoviesList;