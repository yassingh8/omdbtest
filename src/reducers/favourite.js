import {SEARCH_MOVIE, FETCH_MOVIES, BUTTON_CLICK, SHOW_LIST, SHOW_HOME} from '../config/Strings';


const initialState = {
    movieList: [],
    myList: [],
    searchTerm: '',
    showMy: false
}

export default function(state = initialState,action){
    switch(action.type){
        case SEARCH_MOVIE:
            return{
                ...state,
                searchTerm: action.payload,
            }
        case FETCH_MOVIES:
            return{
                ...state,
                movieList: action.payload
            }
        case BUTTON_CLICK:
            return{
                ...state,
                myList: action.payload
            }
        case SHOW_LIST:
            return{
                ...state,
                movieList: action.payload,
                showMy: true
            }
        case SHOW_HOME:
            return{
                ...state,
                movieList: [],
                showMy: false
            }
        default:
            return state
    }
}