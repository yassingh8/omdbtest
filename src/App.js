import React from 'react';
import MovieList from './components/MovieList';
import './App.css';
import store from './store';
import { Provider } from 'react-redux';
import { MdHome } from "react-icons/md";

function App() {
  return (
    <Provider store={store}>
      <MovieList/>
    </Provider>
  );
}

export default App;
