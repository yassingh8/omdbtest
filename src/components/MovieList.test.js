import React from 'react';
// import { render } from '@testing-library/react';
// import ReactDOM from 'react-dom';

import { shallow } from 'enzyme';
import MovieList from './MovieList';
import Movie from './Movie';
import App from '../App';
import configureStore from 'redux-mock-store'; // Smart components
import toJson from 'enzyme-to-json';
import favourite from '../reducers/favourite';
import {SEARCH_MOVIE, FETCH_MOVIES, SHOW_LIST} from '../config/Strings';

const mockStore = configureStore();

const initialState = {
    favourite: {
        movieList: [],
        myList: [],
        searchTerm: '',
        showMy: false
    }
};

const store = mockStore(initialState);

describe('<MovieList />', () => {
    test('renders without crash', () => {
        const wrapper = shallow(<MovieList store={store}/>);
        const movies = wrapper.dive();
        expect(toJson(movies)).toMatchSnapshot();
    });
});

describe('INITIAL_STATE', () => {
    test('is correct', () => {
      const action = { type: 'SEARCH_MOVIE'};
  
      expect(favourite(undefined, action)).toMatchSnapshot();
    });
  })

describe('SHOW_LIST', () => {
    test('is correct', () => {
      const action = { type: 'SHOW_LIST', payload: []};
  
      expect(favourite(undefined, action)).toMatchSnapshot();
    });
  })

// describe('FETCH_MOVIES', () => {
//     test('is correct', () => {
//       const action = { type: 'FETCH_MOVIES'};
  
//       expect(favourite(undefined, action)).toMatchSnapshot();
//     });
//   })
 