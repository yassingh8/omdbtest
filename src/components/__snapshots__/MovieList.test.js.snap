// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MovieList /> renders without crash 1`] = `
<MoviesList
  fetchMovies={[Function]}
  movieList={Array []}
  myList={Array []}
  searchMovie={[Function]}
  searchTerm=""
  showList={[Function]}
  showMy={false}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
/>
`;

exports[`INITIAL_STATE is correct 1`] = `
Object {
  "searchTerm": "",
}
`;

exports[`SHOW_LIST is correct 1`] = `
Object {
  "movieList": Array [],
  "myList": Array [],
  "searchTerm": "",
  "showMy": true,
}
`;
